qualtran/surface_code/fifteen_to_one.py

```diff
@@ -0,0 +1,380 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import lru_cache
+
+import cirq
+import numpy as np
+from attrs import frozen
+
+from qualtran.surface_code.magic_count import MagicCount
+from qualtran.surface_code.magic_state_factory import MagicStateFactory
+from qualtran.surface_code.quantum_error_correction_scheme_summary import (
+    FowlerSuperconductingQubits,
+    QuantumErrorCorrectionSchemeSummary,
+)
+from qualtran.surface_code.t_factory_utils import PauliRotationChannel, storage_error
+
+
+@frozen
+class FifteenToOne(MagicStateFactory):
+    """15-to-1 Magic T state factory.
+
+        reference: https://arxiv.org/abs/1905.06903
+
+    Attributes:
+        d_X: Side length of the surface code along which X measurements happen.
+        d_Z: Side length of the surface code along which Z measurements happen.
+        d_m: Number of code cycles used in lattice surgery.
+        qec: Quantum error correction scheme being used.
+    """
+
+    d_X: int
+    d_Z: int
+    d_m: int
+    qec: QuantumErrorCorrectionSchemeSummary = FowlerSuperconductingQubits
+
+    def footprint(self) -> int:
+        return 2 * (self.d_X + 4 * self.d_Z) * 3 * self.d_X + 4 * self.d_m
+
+    @lru_cache(5)
+    def _final_state(self, phys_err: float):
+        factory = _build_factory(phys_err, self.d_X, self.d_Z, self.d_m, self.qec)
+        return (
+            cirq.DensityMatrixSimulator(dtype=np.complex128).simulate(factory).final_density_matrix
+        )
+
+    @lru_cache(5)
+    def p_fail(self, phys_err: float) -> float:
+        projector = np.kron(np.eye(2), np.ones((16, 16)) / 16)
+        return np.real(1 - np.trace(projector @ self._final_state(phys_err)))
+
+    @lru_cache(5)
+    def p_out(self, phys_err: float) -> float:
+        projector = np.kron(
+            np.eye(2), np.ones((16, 16)) / 16
+        )  # I \otimes ones \otimes ones \otimes ones \otimes ones / 16
```

Comment:
if you put this comment on its own line, the formatter will make this look nicer